### PR TITLE
[Incubator-kie-issues#1605] DMN JIT Executor successfully executing invalid DMN 

### LINKED ISSUE #1605: DMN JIT Executor successfully executing invalid DMN
There is an interesting issue, that once the DMN Runner (JIT Executor) start to execute a DMN model, this model can be updated to be 'invalid', but the runner keeps executing it as it would be a valid one.
Steps to reproduce

    Start BAMOE Canvas with extended services running
    Create simple DMN model
    (Person Age: number) -> (Can Drive?: boolean), where 'Can Drive?' is a literal expression: 'Person Age >= 18'
    Execute such model with inputs: (Person Age: 15) and (Person Age: 30) you will see correct results in the runner form on the right side
    Now remove the edge between (Person Age) and (Can Drive?), the problem tab will inform about a problem
    Try again change the DMN Runner (Person Age) input, it will still continue to evaluate the results [issue]

### PR DESCRIPTION
**Pull request by AthiraHari77**
Fixes https://github.com/apache/incubator-kie-issues/issues/1605

Needs https://github.com/apache/incubator-kie-drools/pull/6146

In this PR, DMN evaluation errors are detected and propagated by throwing an exception upon identification. 
In the event of an exception, a valid status is returned to reflect the error state accurately.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] tests</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-lts</b>
 
- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] native</b>

- for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] native-lts</b>
 
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

<details>
<summary>
Quarkus-3 PR check is failing ... what to do ?
</summary>
The Quarkus 3 check is applying patches from the `.ci/environments/quarkus-3/patches`.

The first patch, called `0001_before_sh.patch`, is generated from Openrewrite `.ci/environments/quarkus-3/quarkus3.yml` recipe. The patch is created to speed up the check. But it may be that some changes in the PR broke this patch.  
No panic, there is an easy way to regenerate it. You just need to comment on the PR:
```
jenkins rewrite quarkus-3
```
and it should, after some minutes (~20/30min) apply a commit on the PR with the patch regenerated.

Other patches were generated manually. If any of it fails, you will need to manually update it... and push your changes.
</details>

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#2` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2142 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2142/2/display/redirect

**Test results:**
- PASSED: 1695
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2142/2/testReport/org.kie.kogito.jobs.embedded/EmbeddedJobsServiceTest/testJobService/">org.kie.kogito.jobs.embedded.EmbeddedJobsServiceTest.testJobService</a></summary>
java.util.ConcurrentModificationException<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1513)<br/>	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)<br/>	at org.kie.kogito.jobs.embedded.EmbeddedJobsServiceTest.testJobService(EmbeddedJobsServiceTest.java:99)<br/>	at java.base/java.lang.reflect.Method.invoke(Method.java:568)<br/>	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1013)<br/>	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:827)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>
</details>

File: jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNResourceHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DMNResourceHelperTest {
+
+    @Test
+    public void testManageResponseWithSuccess() {
+        Supplier<Response> responseSupplier = () -> Response.ok("Success").build();
+        Response response = DMNResourceHelper.manageResponse(responseSupplier);
```
Comment (gitgabrio) [inline jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNResourceHelperTest.java]: Hi @AthiraHari77 
many thanks for the PR!
Could you please implement the auto-closeable try-with-resources here ?

File: jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNResourceHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+        assertEquals("Success", response.getEntity());
+    }
+
+    @Test
+    public void testManageResponseWithFailure() {
+        Supplier<Response> responseSupplier = mock(Supplier.class);
+        when(responseSupplier.get()).thenThrow(new IllegalStateException("Error : Failed to validate"));
+        Response response = DMNResourceHelper.manageResponse(responseSupplier);
```
Comment (gitgabrio) [inline jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNResourceHelperTest.java]: Same comment as above (try-with resources)

File: jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        model = getModelFromIoUtils("invalid_models/DMNv1_x/test.dmn");
+        invalidModel = getModelFromIoUtils("invalid_models/DMNv1_5/DMN-Invalid.dmn");
+        dmnRuntime = mock(DMNRuntime.class);
+        dmnModel = mock(DMNModel.class);
+    }
+
+    @Test
+    void testFromXMLSuccessModel() {
+        String modelXML = model;
+        when(dmnRuntime.getModels()).thenReturn(Collections.singletonList(dmnModel));
```
Comment (gitgabrio) [inline jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java]: What's the need of this mocked 
`when(dmnRuntime.getModels()).thenReturn(Collections.singletonList(dmnModel));` ? 🤔

File: jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private static String model;
+    private static String invalidModel;
+    private static DMNRuntime dmnRuntime;
+    private static DMNModel dmnModel;
+
+    @BeforeAll
+    public static void setup() throws IOException {
+        model = getModelFromIoUtils("invalid_models/DMNv1_x/test.dmn");
+        invalidModel = getModelFromIoUtils("invalid_models/DMNv1_5/DMN-Invalid.dmn");
+        dmnRuntime = mock(DMNRuntime.class);
```
Comment (gitgabrio) [inline jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java]: What's the need of those mocked instances ? 🤔

File: jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        DMNEvaluator evaluator = DMNEvaluator.fromXML(modelXML);
+        assertNotNull(evaluator);
+
+    }
+
+    @Test
+    public void testFromXMLModelWithError() {
+        String modelXML = invalidModel;
+        when(dmnRuntime.getModels()).thenReturn(Collections.singletonList(dmnModel));
```
Comment (gitgabrio) [inline jitexecutor/jitexecutor-dmn/src/test/java/org/kie/kogito/jitexecutor/dmn/DMNEvaluatorTest.java]: Same comment as above

**Review by gitgabrio — COMMENTED** [review]
HI @AthiraHari77 - many thanks for you PR!
I've just some doubts about it

**Review by yesamer — COMMENTED** [review]
Thank you for this PR @AthiraHari77 . I think this change will enforce the error management of the JIT executor and that is great.

My comments:
- Why `JITDMNResource.jitdmn(..)` method doesn't manage the Response as you did in `jitdmnResult` and `jitEvaluateAndExplain`?
- Same question for `SchemaResource` endpoints: why some of them are not managed?

**Comment by AthiraHari77** [comment]
I have fixed the review comments and pushed the changes.
@yesamer I have introduced changes in the **fromXML** method in **DMNEvaluator** to identify the errors and manage exceptions. Hence we are managing the response only in places where the **fromXML** method is invoked.

**Comment by yesamer** [comment]
Thank you @AthiraHari77 for your reply.
I checked and I noticed that **JITDMNResource.jitdmn(..)** indirectly calls **fromXML()** method via **jitdmnService.evaluateModel()**, so I would consider wrapping that endpoint as well to have a consistent behavior over all endpoints. 
About the endpoints in `SchemaResource`, I agree that those don't call  **fromXML()** method, but I would argue that some exceptions may be thrown during the execution, so it would be nice to have the same Response management.
WDTY? :)

**Review by yesamer — APPROVED** [review]
Thank you @AthiraHari77, excellent job!

**Comment by kie-ci3** [comment]
**PR job** `#7` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2142 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2142/7/display/redirect

**Test results:**
- PASSED: 1765
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2142/7/testReport/org.kie.kogito.index/ProcessDataIndexMongoDBIT/testProcessInstanceEvents/">org.kie.kogito.index.ProcessDataIndexMongoDBIT.testProcessInstanceEvents</a></summary>
1 expectation failed.<br/>JSON path data.UserTaskInstances[0].potentialGroups[0] doesn't match.<br/>Expected: managers<br/>  Actual: null<br/>
</details>

**Review by gitgabrio — APPROVED** [review]
Great job @AthiraHari77 !!! LGTM

**Comment by yesamer** [comment]
@gitgabrio Can we merge it?

**Pull request merged**